### PR TITLE
[nuvo] Add NuvoNet source communication capabilities

### PR DESCRIPTION
**Pull request by mlobstein**
Implement NuvoNet source communication that allows for deeper interaction with the physical keypads to include custom menus, custom favorite lists and album art display. The user can configure any/all of the amplifier's 6 sources for use by openHAB. This will allow them to integrate an external source such as a Chromecast to the Nuvo keypads with full metadata and transport controls being passed through openHAB.

The custom menu feature allows the user to create a menu structure that will be available on the physical keypad.  By using rules, it is possible to execute an action on any other openHAB item as a result of selecting a menu item on the physical keypad.

A connection through an MPS4 music server to the Nuvo amplifier must be used in order to utilize these new features. All previous binding functionality available using a serial port connection is unchanged.

### CODE REVIEW

**Comment by mlobstein** [comment]
@boc-tothefuture you might be interested in this

**Comment by boc-tothefuture** [comment]
> @boc-tothefuture you might be interested in this
> 

My MPS4 stopped powering on 2 weeks ago and I loved to direct serial.  

I was not going to repair it, now I might be interested.

**Comment by jlaur** [comment]
@boc-tothefuture - would you like to review this PR?

**Comment by boc-tothefuture** [comment]
@jlaur I no longer have the equipment to test this change (as I say above my MPS4 broke) - but do you just want a code review?

**Comment by jlaur** [comment]
> @jlaur I no longer have the equipment to test this change (as I say above my MPS4 broke) - but do you just want a code review?

Sure, if you are still interested in this binding and could find time for reviewing this PR, that would be nice. :)

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java [hunk 1]
```diff
@@ -70,20 +71,41 @@ public class NuvoBindingConstants {
...
     public static final String BLANK = "";
+    public static final String SPACE = " ";
+    public static final String COMMA = ",";
     public static final String DISPLINE = "DISPLINE";
     public static final String DISPINFO = "DISPINFO,"; // yes comma here
+    public static final String DISPINFOTWO = "DISPINFOTWO";
     public static final String NAME_QUOTE = "NAME\"";
     public static final String MUTE = "MUTE";
     public static final String VOL = "VOL";
+    public static final String OX = "0x";
```
Comment (boc-tothefuture) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java:97]: Can this be changed to a description of what it does?  Is a delimiter, end of line, etc?

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java [hunk 2]
```diff
@@ -70,20 +71,41 @@ public class NuvoBindingConstants {
...
     public static final String OFF = "OFF";
+    public static final String TWO = "2";
     public static final String ONE = "1";
     public static final String ZERO = "0";
     public static final String BLANK = "";
+    public static final String SPACE = " ";
+    public static final String COMMA = ",";
     public static final String DISPLINE = "DISPLINE";
     public static final String DISPINFO = "DISPINFO,"; // yes comma here
+    public static final String DISPINFOTWO = "DISPINFOTWO";
```
Comment (boc-tothefuture) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java:93]: Should this have an underscore between the words?  DISP_INFO_TWO?

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java [hunk 3]
```diff
@@ -70,20 +71,41 @@ public class NuvoBindingConstants {
...
     public static final String MUTE = "MUTE";
     public static final String VOL = "VOL";
+    public static final String OX = "0x";
+    public static final String OCO = "0,0";
 
     // MPS4
     public static final String TYPE_PING = "PING";
+    public static final String TYPE_RESTART = "RESTART";
+    public static final String DISABLE = "disable";
+    public static final String ALBUMARTID = "albumartid";
```
Comment (boc-tothefuture) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java:104]: ALBUM_ART_ID?

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java [hunk 4]
```diff
@@ -70,20 +71,41 @@ public class NuvoBindingConstants {
...
+    public static final String S = "S";
+    public static final String Z = "Z";
+    public static final String ALBUMARTAVAILABLE = "ALBUMARTAVAILABLE";
```
Comment (boc-tothefuture) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java:107]: ALBUM_ART_AVAILABLE

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java [hunk 5]
```diff
@@ -70,20 +71,41 @@ public class NuvoBindingConstants {
...
+    public static final String ALBUMARTFRAG = "ALBUMARTFRAG";
```
Comment (boc-tothefuture) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java:108]: ALBUM_ART_FRAG

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java [hunk 6]
```diff
@@ -70,20 +71,41 @@ public class NuvoBindingConstants {
...
     public static final String VOL = "VOL";
+    public static final String OX = "0x";
+    public static final String OCO = "0,0";
 
     // MPS4
     public static final String TYPE_PING = "PING";
+    public static final String TYPE_RESTART = "RESTART";
+    public static final String DISABLE = "disable";
+    public static final String ALBUMARTID = "albumartid";
+    public static final String S = "S";
```
Comment (boc-tothefuture) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java:105]: Constant for a String that is as long as the constant? 
Can we describe what S and Z are doing here?  What they mean?

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/communication/NuvoImageResizer.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import javax.imageio.ImageIO;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * The {@link NuvoImageResizer} class contains methods for re-sizing album art
+ *
+ * @author Michael Lobstein - Initial contribution
+ */
+@NonNullByDefault
```
Comment (boc-tothefuture) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/communication/NuvoImageResizer.java:33]: I haven't done graphics in java in a LONG LONG time.. I assume the stuff below works in headless mode?

**Comment by boc-tothefuture** [comment]
I put in some minor comments, mostly around constants not being in SCREAMING_SNAKE_CASE to match the others..  Rest of it looks fine to me.

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java (hunk 6 above)
Comment (mlobstein) [reply bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java]: These are just string constants for the source and zone keys. I changed them to be more descriptive.

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java (hunk 1 above)
Comment (mlobstein) [reply bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java]: Changed to be more descriptive.

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java (hunk 2 above)
Comment (mlobstein) [reply bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/NuvoBindingConstants.java]: All variables were changed to SCREAMING_SNAKE_CASE as requested.

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/communication/NuvoImageResizer.java (hunk 7 above)
Comment (mlobstein) [reply bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/communication/NuvoImageResizer.java:33]: Quite well in fact!

**Review by boc-tothefuture — APPROVED** [review]

File: bundles/org.openhab.binding.nuvo/doc/NuvoMenu.xsd [hunk 8]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (fwolter) [inline bundles/org.openhab.binding.nuvo/doc/NuvoMenu.xsd:1]: Is this referenced somewhere?

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java [hunk 9]
```diff
@@ -187,15 +217,65 @@ public void initialize() {
...
         }
 
+        if (this.isMps4) {
+            logger.debug("Port set to {} configuring binding for MPS4 compatability", MPS4_PORT);
+
+            this.isAnyOhNuvoNet = (config.nuvoNetSrc1 == 2 || config.nuvoNetSrc2 == 2 || config.nuvoNetSrc3 == 2
+                    || config.nuvoNetSrc4 == 2 || config.nuvoNetSrc5 == 2 || config.nuvoNetSrc6 == 2);
+
+            if (this.isAnyOhNuvoNet) {
+                logger.debug("At least one source is configured as an OpenHAB NuvoNet source");
```
Comment (fwolter) [inline bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java:233]:
```suggestion
                logger.debug("At least one source is configured as an openHAB NuvoNet source");
```

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java (hunk 9 above)
Comment (mlobstein) [reply bundles/org.openhab.binding.nuvo/src/main/java/org/openhab/binding/nuvo/internal/handler/NuvoHandler.java]: fixed

File: bundles/org.openhab.binding.nuvo/doc/NuvoMenu.xsd (hunk 8 above)
Comment (mlobstein) [reply bundles/org.openhab.binding.nuvo/doc/NuvoMenu.xsd:1]: No it is not. I put this here for a future developer should they wish to regenerate the DTO from the XSD. I just added a note in the file that provides the _'xjc'_ command with all necessary arguments.

**Review by fwolter — APPROVED** [review]
LGTM

**Pull request merged**